### PR TITLE
Device discovery daemon

### PR DESCRIPTION
**Pull request by grunskis**
Converts device discovery into a daemon that's managed by supervisord. Discovery is running every minute and is updating the "DB" with newly discovered devices. Frontend is polling the GET endpoint for new devices. It's also possible to restart the discovery by POSTing to `/-/setup/devices/discover` which will trigger a restart of supervisor program through their RPC interface.

`discover_devices` command doesn't have any tests, but probably should. Let me know what you think...

Includes also other fixes and improvements that I've stumbled upon while working on this. Sorry about that.

### CODE REVIEW

File: application/senic/nuimo_hub/commands.py [hunk 1]
```diff
@@ -228,3 +242,55 @@ def get_component_for_device(device):
...
+        devices = []
+
+    while True:
+        now = datetime.utcnow()
+
+        # remove discovery time from known devices, to simplify logic
+        # of finding only the new devices
+        known_devices = deepcopy(devices)
+        for device in known_devices:
+            device.pop("timeOfDiscovery", None)
```
Comment (larsblumberg) [inline application/senic/nuimo_hub/commands.py]: `discoveryDate` or `discoveryTimestamp` as an easier name?

File: deployment/roles/hub-ubuntu/tasks/main.yml [hunk 2]
```diff
@@ -137,6 +147,7 @@
   with_items:
     - scan_wifi
     - wifi_setup
+    - discover_devices
```
Comment (larsblumberg) [inline deployment/roles/hub-ubuntu/tasks/main.yml]: If would opt for using nouns for command line tools, i.e. `device_discovery`

File: frontend/src/SetupDevices.js [hunk 3]
```diff
@@ -5,6 +5,9 @@ import { Link } from 'react-router'
 import './SetupDevices.css'
 
 class SetupDevices extends Component {
+  devicePollInterval = 5000  // 5 seconds
+  devicePollTimer = null
```
Comment (larsblumberg) [inline frontend/src/SetupDevices.js]: Since we're polling for multiple devices, can we name it `devicesPollInterval` and `devicesPollTimer`?

File: application/senic/nuimo_hub/commands.py [hunk 4]
```diff
@@ -228,3 +242,55 @@ def get_component_for_device(device):
...
+
+        discovered_devices = discover(known_devices=known_devices)
+
+        if discovered_devices:
+            for device in discovered_devices:
```
Comment (larsblumberg) [inline application/senic/nuimo_hub/commands.py]: To avoid `if` and thus one nesting level, we could type:

    for device in (discovered_devices or [])

File: application/senic/nuimo_hub/commands.py [hunk 5]
```diff
@@ -228,3 +242,55 @@ def get_component_for_device(device):
...
+                # make sure we get updates for devices we already had discovered before
+                devices = [x for x in devices if x["id"] != device["id"]]
```
Comment (larsblumberg) [inline application/senic/nuimo_hub/commands.py]: I'm fine with one letter vars in comprehensions but can we use the initial letter?, i.e. `for d in devices`

File: application/senic/nuimo_hub/commands.py [hunk 6]
```diff
@@ -228,3 +242,55 @@ def get_component_for_device(device):
...
+                device["timeOfDiscovery"] = str(now)
+                devices.append(device)
+
+            with open(devices_path, 'w') as f:
+                json.dump(devices, f)
+
+        next_scan = now + timedelta(seconds=scan_interval_seconds)
+        logging.info("Next device discovery run scheduled for {}".format(next_scan))
+        while datetime.utcnow() < next_scan:
+            time.sleep(1)
```
Comment (larsblumberg) [inline application/senic/nuimo_hub/commands.py]: Alternatively, can we sleep the remaining time without using `while`? I.e.

    time.sleep(next_scan - datetime.utcnow())

File: application/senic/nuimo_hub/device_discovery.py [hunk 7]
```diff
@@ -35,7 +36,16 @@ class PhilipsHueBridgeError(IntEnum):
 logger = logging.getLogger(__name__)
 
 
-def discover(discovery_class=NetworkDiscovery):
+def discover(discovery_class=NetworkDiscovery, known_devices=None):
```
Comment (larsblumberg) [inline application/senic/nuimo_hub/device_discovery.py]: It's better do default `known_devices` to `[]` and not allow `None`, then we don't need to test for `known_devices is None` as you do later

File: application/senic/nuimo_hub/device_discovery.py [hunk 8]
```diff
@@ -50,6 +60,10 @@ def discover(discovery_class=NetworkDiscovery):
         for device_info in netdisco.get_info(device_type):
             device_description = make_device_description(device_type, device_info)
             if device_description:
+                if known_devices and device_description in known_devices:
+                    continue
+
+                logger.info("Discovered new %s device with ip %s", device_type, device_description["ip"])
                 devices.append(device_description)
```
Comment (larsblumberg) [inline application/senic/nuimo_hub/device_discovery.py]:
```python
    device_descriptions = [make_device_description(device_type, d) for d in netdisco.get_info(device_type)]
    new_device_descriptions = [d for d in device_descriptions if d not in known_devices]
    logger.info(...)
    devices.extend(new_device_descriptions)
```

File: deployment/roles/hub-ubuntu/templates/discover_devices_supervisor.conf [hunk 9]
```diff
@@ -0,0 +1,10 @@
+[program:discover_devices]
+command = {{backend_deploy_location}}/venv/bin/discover_devices -c {{backend_deploy_location}}/production.ini
+autostart=true
+autorestart=true
+directory={{ backend_deploy_location }}
+stdout_logfile={{ backend_data_location }}/discover_devices.log
+redirect_stderr=true
+user={{ run_user }}
+startsecs=3
+environment=LC_ALL=C.UTF-8, LANG=C.UTF-8
```
Comment (larsblumberg) [inline deployment/roles/hub-ubuntu/templates/discover_devices_supervisor.conf]: If you rename the cmd line tool to `device_discovery`, this .conf file also needs to be renamed

File: nuimo_app/senic/nuimo_app/__main__.py [hunk 10]
```diff
@@ -24,24 +24,24 @@ def main(config_file_path=DEFAULT_CONFIG_FILE_PATH):
...
+    ha_url = config.get("ha_api_url", "localhost:8123")
     ha_api = HAListener("ws://{}".format(ha_url))
     ha_api.start()
 
     nuimo_app = None
     controller_mac_address = config.get("controller_mac_address")
     if controller_mac_address:
-        ble_adapter_name = config.get("ble_adapter_name", DEFAULT_BLE_ADAPTER_NAME)
-        nuimo_app = NuimoApp(controller_mac_address, ha_api, ble_adapter_name)
+        nuimo_app = NuimoApp(ha_api, controller_mac_address, manager)
```
Comment (larsblumberg) [inline nuimo_app/senic/nuimo_app/__main__.py]: Shouldn't we let `NuimoApp` create the `ControllerManager` internally and continue to pass `ble_adapter_name` to `NuimoApp()`?

File: application/senic/nuimo_hub/commands.py (hunk 6 above)
Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: Wouldn't that block the main thread?

Comment (larsblumberg) [inline application/senic/nuimo_hub/commands.py]: Doesn't it block it anyway already?

File: nuimo_app/senic/nuimo_app/__main__.py (hunk 10 above)
Comment (grunskis) [inline nuimo_app/senic/nuimo_app/__main__.py]: Maybe that could work. I remember that there was a problem that Ctrl+C wasn't responding otherwise, but that might have changed. I'll look into this...

File: application/senic/nuimo_hub/commands.py (hunk 4 above)
Comment (larsblumberg) [inline application/senic/nuimo_hub/commands.py]: Alternatively:

    discovered_devices = discover(known_devices=known_devices) or []

Comment (larsblumberg) [inline application/senic/nuimo_hub/commands.py]: And alternatively, even better :) `discover` method never returns `None`, always an array (or dict, dunno)

File: application/senic/nuimo_hub/commands.py (hunk 5 above)
Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: `x` is the standard name. like `i` in C for looping. IMHO it also looks better in most cases than any other letter :D

File: application/senic/nuimo_hub/commands.py (hunk 6 above)
Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: Nope, it doesn't. If it would then supervisor wouldn't be able to restart the program... I'll check how it behaves with long sleeps...

Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: Yep, killing still works also when sleeping the whole period, fa3313e

File: frontend/src/SetupDevices.js (hunk 3 above)
Comment (grunskis) [inline frontend/src/SetupDevices.js]: Renamed b26aa64

File: application/senic/nuimo_hub/commands.py (hunk 1 above)
Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: Renamed afd5342

File: application/senic/nuimo_hub/commands.py (hunk 4 above)
Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: It actueally doesn't return `None`

File: application/senic/nuimo_hub/commands.py (hunk 5 above)
Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: Refactored and took care of this. By popular demand will avoid using `x` as var name ;)

File: application/senic/nuimo_hub/commands.py (hunk 1 above)
Comment (grunskis) [inline application/senic/nuimo_hub/commands.py]: In the end renamed the field to `discovered` to match naming scheme of other fields which don't include type as part of name

File: nuimo_app/senic/nuimo_app/__main__.py (hunk 10 above)
Comment (grunskis) [inline nuimo_app/senic/nuimo_app/__main__.py]: Did as you suggested here 0c59d6d Thanks!

**Comment by grunskis** [comment]
@larsblumberg addressed all the things besides rename of the daemon. if we decide to rename it then we should use the same naming pattern for all daemons/programs that we have.

Let me know if you have any other comments, otherwise I'll squash...

**Comment by larsblumberg** [comment]
LGTM! Nice PR!
